framework/src/org/checkerframework/framework/util/ViewpointAdaptor.java

```diff
@@ -0,0 +1,375 @@
+     * @param lhs type from which type arguments are extracted to replace formal type parameters of
+     *     rhs.
+     * @param rhs AnnotatedTypeMirror that might be a formal type parameter
+     * @return rhs' copy with its type parameter substituted
+     */
+    private AnnotatedTypeMirror substituteTVars(
+            AnnotatedTypeFactory f, AnnotatedTypeMirror lhs, AnnotatedTypeMirror rhs) {
+        if (rhs.getKind() == TypeKind.TYPEVAR) {
+            AnnotatedTypeVariable atv = (AnnotatedTypeVariable) rhs.shallowCopy();
+
+            // Base case where actual type argument is extracted
+            if (lhs.getKind() == TypeKind.DECLARED) {
+                rhs = getTypeVariableSubstitution(f, (AnnotatedDeclaredType) lhs, atv);
+            }
+            // else TODO: the receiver might be another type variable... should we do something?
+        } else if (rhs.getKind() == TypeKind.DECLARED) {
+            //System.out.println("before: " + rhs);
+            AnnotatedDeclaredType adt = (AnnotatedDeclaredType) rhs.shallowCopy();
+            Map<AnnotatedTypeMirror, AnnotatedTypeMirror> mapping = new HashMap<>();
+
+            for (AnnotatedTypeMirror formalTypeParameter : adt.getTypeArguments()) {
+                AnnotatedTypeMirror actualTypeArgument =
+                        substituteTVars(f, lhs, formalTypeParameter);
+                mapping.put(formalTypeParameter, actualTypeArgument);
+                // The following code does the wrong thing!
+            }
+            // We must use AnnotatedTypeReplacer to replace the formal type parameters with actual type
+            // arguments, but not replace with its main modifier
+            rhs = AnnotatedTypeReplacer.replace(adt, mapping);
+        } else if (rhs.getKind() == TypeKind.WILDCARD) {
+            AnnotatedWildcardType awt = (AnnotatedWildcardType) rhs.shallowCopy();
+            Map<AnnotatedTypeMirror, AnnotatedTypeMirror> mapping = new HashMap<>();
+
+            AnnotatedTypeMirror extend = awt.getExtendsBound();
+            if (extend != null) {
+                AnnotatedTypeMirror substExtend = substituteTVars(f, lhs, extend);
+                mapping.put(extend, substExtend);
+            }
+
+            AnnotatedTypeMirror zuper = awt.getSuperBound();
+            if (zuper != null) {
+                AnnotatedTypeMirror substZuper = substituteTVars(f, lhs, zuper);
+                mapping.put(zuper, substZuper);
+            }
+
+            rhs = AnnotatedTypeReplacer.replace(awt, mapping);
+        } else if (rhs.getKind() == TypeKind.ARRAY) {
+            AnnotatedArrayType aat = (AnnotatedArrayType) rhs.shallowCopy();
+            Map<AnnotatedTypeMirror, AnnotatedTypeMirror> mapping = new HashMap<>();
+
+            AnnotatedTypeMirror compnentType = aat.getComponentType();
+            // Type variable of compnentType already gets substituted
+            AnnotatedTypeMirror substCompnentType = substituteTVars(f, lhs, compnentType);
```

Comment:
I  think both answers are no. They both substitute only one level. I think the "recursive" process is handled by visitors. So this operation only handles one level.